io-sim-classes/src/Control/Monad/Class/MonadSTM/Strict.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+module Control.Monad.Class.MonadSTM.Strict
+  ( module X
+    -- * 'StrictTVar'
+  , StrictTVar
+  , newTVar
+  , newTVarM
+  , newTVarWithInvariantM
+  , readTVar
+  , writeTVar
+  , modifyTVar
+  , updateTVar
+    -- * 'StrictTMVar'
+  , StrictTMVar
+  , newTMVar
+  , newTMVarM
+  , newTMVarWithInvariantM
+  , newEmptyTMVar
+  , newEmptyTMVarM
+  , newEmptyTMVarWithInvariantM
+  , takeTMVar
+  , tryTakeTMVar
+  , putTMVar
+  , tryPutTMVar
+  , readTMVar
+  , tryReadTMVar
+  , swapTMVar
+  , isEmptyTMVar
+  ) where
+
+import           Control.Exception (assert)
+import           Control.Monad.Class.MonadSTM as X hiding (LazyTMVar, LazyTVar,
+                     isEmptyTMVar, modifyTVar, newEmptyTMVar, newEmptyTMVarM,
+                     newTMVar, newTMVarM, newTVar, newTVarM, putTMVar,
+                     readTMVar, readTVar, swapTMVar, takeTMVar, tryPutTMVar,
+                     tryReadTMVar, tryTakeTMVar, updateTVar, writeTVar)
+import qualified Control.Monad.Class.MonadSTM as Lazy
+
+{-------------------------------------------------------------------------------
+  Strict TVar
+-------------------------------------------------------------------------------}
+
+data StrictTVar m a = StrictTVar
+   { invariant :: !(a -> Bool)
+     -- ^ Invariant checked in an 'assert' whenever storing an @a@ in the
+     -- 'StrictTVar'.
+   , tvar      :: !(Lazy.LazyTVar m a)
```

Comment:
Not having looked at the PR yet, but just to answer the question: yes, the `StrictTVar` should ensure that its value is in WHNF at all times (new, update, set, what have you). The idea behind the invariant is that we can then plug in your "is this thing in NF?" check, @intricate , so that we can verify it's not only in WHNF, but indeed in NF.